src/sycltest/bin/main.cc

```diff
@@ -0,0 +1,150 @@
+      print_help(args.front());
+      return EXIT_FAILURE;
+    }
+  }
+  if (numberOfStreams == 0) {
+    numberOfStreams = numberOfThreads;
+  }
+  if (datadir.empty()) {
+    datadir = std::filesystem::path(args[0]).parent_path() / "data";
+  }
+  if (not std::filesystem::exists(datadir)) {
+    std::cout << "Data directory '" << datadir << "' does not exist" << std::endl;
+    return EXIT_FAILURE;
+  }
+
+  // Initialse the CUDA runtime
+  cms::sycltools::enumerateDevices(true);
+
+  // Initialize EventProcessor
+  std::vector<std::string> edmodules;
+  std::vector<std::string> esmodules;
+  if (not empty) {
+    edmodules = {"TestProducer", "TestProducer3", "TestProducer2"};
+    esmodules = {"IntESProducer"};
+    if (transfer) {
+      // add modules for transfer
+    }
+  }
+  edm::EventProcessor processor(
+      maxEvents, numberOfStreams, std::move(edmodules), std::move(esmodules), datadir, validation);
+  maxEvents = processor.maxEvents();
+
+  std::cout << "Processing " << maxEvents << " events, of which " << numberOfStreams << " concurrently, with "
+            << numberOfThreads << " threads." << std::endl;
+
+  // Initialize tasks scheduler (thread pool)
+  tbb::task_scheduler_init tsi(numberOfThreads);
+
+  // Run work
+  auto start = std::chrono::high_resolution_clock::now();
+  try {
+    processor.runToCompletion();
+  } catch (std::runtime_error& e) {
+    std::cout << "\n----------\nCaught std::runtime_error" << std::endl;
+    std::cout << e.what() << std::endl;
+    return EXIT_FAILURE;
+  } catch (std::exception& e) {
+    std::cout << "\n----------\nCaught std::exception" << std::endl;
+    std::cout << e.what() << std::endl;
+    return EXIT_FAILURE;
+  } catch (...) {
+    std::cout << "\n----------\nCaught exception of unknown type" << std::endl;
+    return EXIT_FAILURE;
+  }
+  auto stop = std::chrono::high_resolution_clock::now();
+
+  // Run endJob
+  try {
+    processor.endJob();
+  } catch (std::runtime_error& e) {
+    std::cout << "\n----------\nCaught std::runtime_error" << std::endl;
+    std::cout << e.what() << std::endl;
+    return EXIT_FAILURE;
+  } catch (std::exception& e) {
+    std::cout << "\n----------\nCaught std::exception" << std::endl;
+    std::cout << e.what() << std::endl;
+    return EXIT_FAILURE;
+  } catch (...) {
+    std::cout << "\n----------\nCaught exception of unknown type" << std::endl;
+    return EXIT_FAILURE;
+  }
+
+  // Work done, report timing
+  auto diff = stop - start;
+  auto time = static_cast<double>(std::chrono::duration_cast<std::chrono::microseconds>(diff).count()) / 1e6;
+  std::cout << "Processed " << maxEvents << " events in " << std::scientific << time << " seconds, throughput "
+            << std::defaultfloat << (maxEvents / time) << " events/s." << std::endl;
+  return EXIT_SUCCESS;
+}
+catch (sycl::exception const& exc) {
+  std::cerr << exc.what() << "Exception caught at file:" << __FILE__ << ", line:" << __LINE__ << std::endl;
+  std::exit(1);
+}
```

Comment:
> Ok for now, but could be the try-catch scope be smaller?

This was added automatically by the migration - I don't know if because we call some SYCL functions from `main`, or a priori to make sure all SYCL exceptions are handled.

In CMSSW we can rely on the framework catching all exceptions, so we don't need to have such overarching `catch` block.

> I suppose in CMSSW the `cms::sycltools::enumerateDevices(true);` (and possible other initialization) would go into a Service.

Yes, same as for the call to `cudaGetDeviceCount(&numberOfDevices);` in `cudatest` *vs* using `CUDAService` in CMSSW.